lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+import tarfile
+import tempfile
+import time
+import uuid
+import yaml
+
+from contextlib import contextmanager
+from distutils.version import LooseVersion
+from hashlib import sha256
+from io import BytesIO
+from yaml.error import YAMLError
+
+try:
+    from urllib.parse import urlparse  # Python 3
+except ImportError:
+    from urlparse import urlparse  # Python 2
+
+import ansible.constants as C
+import ansible.module_utils.six.moves.urllib.error as urllib_error
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_bytes, to_native, to_text
+from ansible.utils.display import Display
+from ansible.utils.hashing import secure_hash, secure_hash_s
+
+from ansible.module_utils.urls import open_url
+
+
+display = Display()
+
+MANIFEST_FORMAT = 1
+
+
+class CollectionRequirement(object):
+
+    _FILE_MAPPING = [('MANIFEST.json', 'manifest'), ('FILES.json', 'files')]
+
+    def __init__(self, namespace, name, path, source, versions, requirement, force, parent=None, validate_certs=True,
+                 metadata=None, files=None, skip=False):
+        """
+        Represents a collection requirement, the versions that are available to be installed as well as any
+        dependencies the collection has.
+
+        :param namespace: The collection namespace.
+        :param name: The collection name.
+        :param path: The path to the collection tarball if it has already been downloaded.
+        :param source: The Galaxy server URL to download if the collection from Galaxy.
+        :param versions: A list of versions of the collection that is available
+        :param requirement: The version requirement string used to verify the list of versions fit the requirements.
+        :param force: Whether the force flag applied to the collection.
+        :param parent: The name of the parent the collection is a dependency of.
+        :param validate_certs: Whether to validate the Galaxy server certificate.
+        :param metadata: The collection metadata dict if it has already been retrieved.
+        :param files: The files that exist inside the collection. This is based on the FILES.json file inside the
+            collection artifact.
+        :param skip: Whether to skip installing the collection, should be set if the collection is already installed
+            and force is not set.
+        """
+        self.namespace = namespace
+        self.name = name
+        self.path = path
+        self.source = source
+        self.versions = set(versions)
+        self.force = force
+        self.skip = skip
+        self.required_by = []
+        self._validate_certs = validate_certs
+
+        self._metadata = metadata
+        self._files = files
+        self._galaxy_info = None
+
+        self.add_requirement(parent, requirement)
+
+    def __str__(self):
+        return "%s.%s" % (self.namespace, self.name)
+
+    @property
+    def latest_version(self):
+        versions = [v for v in self.versions if v != '*']
```

Comment:
You don't need a copy of the list in memory here. Better use a generator expression:
```suggestion
        versions = (v for v in self.versions if v != '*')
```